rust/src/fvm/machine.rs

```diff
@@ -0,0 +1,264 @@
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use std::mem;
+use std::ptr;
+use std::sync::{atomic::AtomicU64, Mutex};
+
+use anyhow::Error;
+use blockstore::cgo::CgoBlockstore;
+use blockstore::{Block, Blockstore, MemoryBlockstore};
+use cid::Cid;
+use drop_struct_macro_derive::DropStructMacro;
+use ffi_toolkit::{
+    c_str_to_pbuf, catch_panic_response, raw_ptr, rust_str_to_c_str, FCPResponseStatus,
+};
+use fvm::externs::cgo::CgoExterns;
+use fvm::externs::Externs;
+use fvm::machine::{ApplyKind, ApplyRet, Machine};
+use fvm::message::Message;
+use fvm::Config;
+use fvm_shared::address::Address;
+use fvm_shared::clock::ChainEpoch;
+use fvm_shared::econ::TokenAmount;
+use fvm_shared::encoding::RawBytes;
+use fvm_shared::version::NetworkVersion;
+use fvm_shared::MethodNum;
+use log::{error, info};
+use num_traits::FromPrimitive;
+use once_cell::sync::Lazy;
+
+use super::types::*;
+use crate::util::api::init_log;
+
+type CgoMachine = Machine<CgoBlockstore, CgoExterns>;
+
+static FVM_MAP: Lazy<Mutex<HashMap<u64, CgoMachine>>> =
+    Lazy::new(|| Mutex::new(HashMap::with_capacity(1)));
+
+const NEXT_ID: AtomicU64 = AtomicU64::new(0);
+
+fn add_fvm_machine(machine: CgoMachine) -> u64 {
+    let next_id = NEXT_ID.fetch_add(1, std::sync::atomic::Ordering::SeqCst);
+    let mut machines = FVM_MAP.lock().unwrap();
+    machines.insert(next_id, machine);
+    next_id
+}
+
+fn get_default_config() -> fvm::Config {
+    Config {
+        initial_pages: 1024, //FIXME
+        max_pages: 32768,    // FIXME
+        engine: wasmtime::Config::new(),
+    }
+}
+
+/// Note: the incoming args as u64 and odd conversions to i32/i64
+/// for some types is due to the generated bindings not liking the
+/// 32bit types as incoming args
+///
+#[no_mangle]
+#[cfg(not(target_os = "windows"))]
+pub unsafe extern "C" fn fil_create_fvm_machine(
+    fvm_version: fil_FvmRegisteredVersion,
+    chain_epoch: u64,
+    token_amount: u64,
```

Comment:
Is this the basefee? Token amounts are BigInts, [effectively bounded at 91 bits](https://github.com/filecoin-project/fvm/issues/56) given the current maximum supply in AttoFiL (unit of exchange).